### ISSUE TITLE
Fetching null linked-resource throws exception

### ISSUE THREAD

**Issue by vincbon** [BlackPepperSoftware/bowman#33]
Hello,

We're currently using Bowman with SDR and have an issue regarding nullable linked resources.

Bowman throws an exception when trying to fetch a null linked-resource because the link is effectively absent. We have to wrap affected getters around try-catches, reducing the great transparency that Bowman initially offers.

This is not the case for inline resources as the property is set to "null" in the json.

I'm not sure if this is the correct behavior. It helps debugging typos but from a semantic point of view the absence of a link just denotes the absence of an association. Thus it may be better to return null if the link cannot be found ?

What's your opinion on this ?

**Comment by hdpe** [comment]
Thanks very much for raising this. I've been pondering this over the weekend. I think you're right, and can totally see how this behaviour would be desirable.

It's always a delicate balance adding new convenience yet ensuring debuggability following user error, while not breaking client code where possible and trying to maintain a consistent API and the (sometimes tenuous) JPA analogy.

In JPA, `@ManyToOne` etc. can be marked `optional` (the default) to mean that there doesn't need to be a value on the other side of the association. But if the association doesn't exist *at all*, e.g. you've tried to follow a join that doesn't exist, it will error rather than return null or an empty collection.

I think maintaining this in Bowman is even more important once you're looking at things like templated links - if a link like `http://localhost:8080/x/fndByName{?name}` silently returned null just because the `fndByName` rel was misspelt and not found, I think that mistake could be really hard to spot.

So... I think it should be opt-in. How about:

```
@LinkedResource(optionalLink = true) public X getX() { ... }
```

where `false` would be the default, so you have to explicitly opt-in to reduced error reporting and possibly breaking your existing error handling. `optionalLink` seems clearer to me than e.g. `optional` to make explicit that it's the *link* that is absent, rather than the *value* when that link is retrieved.

What do you think? The downside, as you say, is that this isn't consistent with how inline resources work, but I think that may be a price worth paying.

**Comment by vincbon** [comment]
This seems reasonable to me, I did not think of an optional flag. You're right returning null everytime is probably too loose and error-prone.

In our case the nullable relations are identified beforehand so an opt-in mecanism is totally fine.

Thanks again for this awesome library ! It really eases the development of an HAL client when you're used to JPA.

**Comment by hdpe** [comment]
Implemented and released as `0.8.0`. Thanks for raising this and for your kind words about Bowman. :)
